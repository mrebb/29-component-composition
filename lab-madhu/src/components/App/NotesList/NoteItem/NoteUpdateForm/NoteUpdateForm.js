import React, { Component, Fragment } from 'react';
export default class NoteUpdateForm extends Component {
  constructor(props) {
    super(props);
  
    this.state = {
      editing: this.props.note.editing,
      completed: this.props.note.completed,
      content: this.props.note.content,
      title: this.props.note.title,
    };
  }
  onUpdate = event => {
    event.preventDefault();
    this.props.onUpdate({ ...this.state, id: this.props.note.id });
    this.props.showComponentMethod(false);
  };
  onChange = event => {
    let changedVal =
      event.target.type === 'checkbox'
        ? event.target.checked
        : event.target.value;
    const changedBit = { [event.target.name]: changedVal };
    this.setState(changedBit);
  };
  onCancel = event => {
    event.preventDefault();
    this.props.showComponentMethod(false);
  };
  render() {
    return (
      <Fragment>
        <form
          className="updateForm"
          onSubmit={this.onUpdate}
          onChange={this.onChange}
        >
          <input
            name="title"
            defaultValue={this.props.note.title}
            type="text"
            placeholder="title"
          />
          <textarea
            name="content"
            defaultValue={this.props.note.content}
            placeholder="content"
          />
          <label>
            <span>editing</span>
            <input
              defaultChecked={this.props.note.editing}
              name="editing"
              type="checkbox"
            />
          </label>
          <label>
            <span>completed</span>
          </label>
          <input
            defaultChecked={this.props.note.completed}
            name="completed"
            type="checkbox"
          />
          <button>update note</button>
          <button onClick={this.onCancel}>cancel</button>
        </form>
      </Fragment>
    );
  }
}
