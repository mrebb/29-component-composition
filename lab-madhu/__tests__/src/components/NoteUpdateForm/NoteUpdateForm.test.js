import React from 'react';
import renderer from 'react-test-renderer';
import { shallow,mount } from 'enzyme';
import NoteUpdateForm from '../../../../src/components/App/NotesList/NoteItem/NoteUpdateForm/NoteUpdateForm';

describe('NoteUpdateForm component', () => {
  it('should have initial state', () => {
    const note = {
      editing: false,
      completed: false,
      content: 'sample content',
      title: 'test title',
      id:1234,
    };
    const wrapper = mount(<NoteUpdateForm note={note}/>);
    expect(wrapper).toHaveState({
      editing: false,
      completed: false,
      content: 'sample content',
      title: 'test title',
    });
  });
  it('should handle form changes', () => {
    const note = {
      editing: false,
      completed: false,
      content: 'sample content',
      title: 'test title',
    };
    const wrapper = mount(<NoteUpdateForm note={note}/>);
    const instance = wrapper.instance();
    const event = {
      target: {
        name: 'title',
        value: 'Hello world',
        type: 'text',
      },
    };
    instance.onChange(event);

    expect(wrapper.state('title')).toBe('Hello world');
  });

  it('should update existing note', done => {
    const note = {
      editing: false,
      completed: false,
      content: 'sample content',
      title: 'test title',
      id:1234,
    };
    const onUpdate = note => {
      expect(note.id).toBe(1234);
      expect(note.title).toBe('Hello world');
      done();
    };
    const wrapper = mount(<NoteUpdateForm onUpdate={onUpdate} showComponentMethod={()=>{return true;}}note={note}/>);
    const instance = wrapper.instance();
    const e = {
      target: {
        name: 'title',
        value: 'Hello world',
        type: 'text',
      },
    };
    instance.onChange(e);
    const event = {
      preventDefault: () => {},
    };

    instance.onUpdate(event);
  });
});

describe('<NoteUpdateForm/> (Snapshot Test)', () => {
  it('renders right', () => {
    const note = {
      editing: false,
      completed: false,
      content: 'sample content',
      title: 'test title',
    };
    const component = renderer.create(
      <NoteUpdateForm note={note}/>
    );

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();

  });
});
