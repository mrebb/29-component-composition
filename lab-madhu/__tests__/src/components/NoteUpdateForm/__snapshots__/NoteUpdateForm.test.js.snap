// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteUpdateForm/> (Snapshot Test) renders right 1`] = `
<form
  className="updateForm"
  onChange={[Function]}
  onSubmit={[Function]}
>
  <input
    defaultValue="test title"
    name="title"
    placeholder="title"
    type="text"
  />
  <textarea
    defaultValue="sample content"
    name="content"
    placeholder="content"
  />
  <label>
    <span>
      editing
    </span>
    <input
      defaultChecked={false}
      name="editing"
      type="checkbox"
    />
  </label>
  <label>
    <span>
      completed
    </span>
  </label>
  <input
    defaultChecked={false}
    name="completed"
    type="checkbox"
  />
  <button>
    update note
  </button>
  <button
    onClick={[Function]}
  >
    cancel
  </button>
</form>
`;
